File: api/OnlineUpdate.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// OnlineUpdate 对应 OnlineUpdate.py。
type OnlineUpdate struct {
	*session.ApiGroup
}

// NewOnlineUpdate 创建 OnlineUpdate API 分组。
func NewOnlineUpdate(s *session.Session) *OnlineUpdate {
	return &OnlineUpdate{ApiGroup: session.NewApiGroup(s)}
}

// CheckNewVersion 对应 online-update/check-new-version。
func (o *OnlineUpdate) CheckNewVersion() (map[string]interface{}, error) {
	return o.S.Get("online-update/check-new-version", nil, "api")
}

// SetCheckNewVersion 对应 online-update/check-new-version (post_set, 无数据)。
func (o *OnlineUpdate) SetCheckNewVersion() (interface{}, error) {
	return o.S.PostSet("online-update/check-new-version", nil, false, "api", false, false)
}

// Status 对应 online-update/status。
func (o *OnlineUpdate) Status() (map[string]interface{}, error) {
	return o.S.Get("online-update/status", nil, "api")
}

// UrlList 对应 online-update/url-list。
func (o *OnlineUpdate) UrlList() (map[string]interface{}, error) {
	return o.S.Get("online-update/url-list", nil, "api")
}

// AckNewversion 对应 online-update/ack-newversion。
func (o *OnlineUpdate) AckNewversion() (map[string]interface{}, error) {
	return o.S.Get("online-update/ack-newversion", nil, "api")
}

// SetAckNewversion 对应 online-update/ack-newversion (post_set)。
func (o *OnlineUpdate) SetAckNewversion() (interface{}, error) {
	return o.S.PostSet("online-update/ack-newversion", map[string]interface{}{
		"userAckNewVersion": 0,
	}, false, "api", false, false)
}

// CancelDownloading 对应 online-update/cancel-downloading。
func (o *OnlineUpdate) CancelDownloading() (map[string]interface{}, error) {
	return o.S.Get("online-update/cancel-downloading", nil, "api")
}

// SetCancelDownloading 对应 online-update/cancel-downloading (post_set, 无数据)。
func (o *OnlineUpdate) SetCancelDownloading() (interface{}, error) {
	return o.S.PostSet("online-update/cancel-downloading", nil, false, "api", false, false)
}

// UpgradeMessagebox 对应 online-update/upgrade-messagebox。
func (o *OnlineUpdate) UpgradeMessagebox() (map[string]interface{}, error) {
	return o.S.Get("online-update/upgrade-messagebox", nil, "api")
}

// SetUpgradeMessagebox 对应 online-update/upgrade-messagebox (post_set)。
func (o *OnlineUpdate) SetUpgradeMessagebox(messagebox string) (interface{}, error) {
	return o.S.PostSet("online-update/upgrade-messagebox", map[string]interface{}{
		"messagebox": messagebox,
	}, false, "api", false, false)
}

// Configuration 对应 online-update/configuration。
func (o *OnlineUpdate) Configuration() (map[string]interface{}, error) {
	return o.S.Get("online-update/configuration", nil, "api")
}

// AutoupdateConfig 对应 online-update/autoupdate-config。
func (o *OnlineUpdate) AutoupdateConfig() (map[string]interface{}, error) {
	return o.S.Get("online-update/autoupdate-config", nil, "api")
}

// SetAutoupdateConfig 对应 online-update/autoupdate-config (post_set)。
func (o *OnlineUpdate) SetAutoupdateConfig(autoupdate bool) (interface{}, error) {
	val := 0
	if autoupdate {
		val = 1
	}
	return o.S.PostSet("online-update/autoupdate-config", map[string]interface{}{
		"auto_update": val,
		"ui_download": 0,
	}, false, "api", false, false)
}

// RedirectCancel 对应 online-update/redirect_cancel。
func (o *OnlineUpdate) RedirectCancel() (map[string]interface{}, error) {
	return o.S.Get("online-update/redirect_cancel", nil, "api")
}

File: src/api/online_update.rs
//! OnlineUpdate API group (`api/OnlineUpdate.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::XmlMap;

use super::{get_value, post_set};

/// OnlineUpdate API group.
pub struct OnlineUpdate<'a> {
    conn: &'a Connection,
}

impl<'a> OnlineUpdate<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        OnlineUpdate { conn }
    }

    /// `online-update/check-new-version`. Check for a new firmware version.
    pub fn check_new_version(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/check-new-version")
    }

    /// Trigger a check for a new firmware version.
    pub fn set_check_new_version(&self) -> Result<String> {
        let body = XmlMap::new();
        post_set(self.conn, "api/online-update/check-new-version", &body)
    }

    /// `online-update/status`. Online update status.
    pub fn status(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/status")
    }

    /// `online-update/url-list`. List of URLs for the online update.
    pub fn url_list(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/url-list")
    }

    /// `online-update/ack-newversion`. Acknowledge the new firmware version.
    pub fn ack_newversion(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/ack-newversion")
    }

    /// Acknowledge the new firmware version.
    pub fn set_ack_newversion(&self) -> Result<String> {
        let body: XmlMap = map_of([("userAckNewVersion", "0".to_string())]);
        post_set(self.conn, "api/online-update/ack-newversion", &body)
    }

    /// `online-update/cancel-downloading`. Cancel the firmware download.
    pub fn cancel_downloading(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/cancel-downloading")
    }

    /// Cancel the firmware download.
    pub fn set_cancel_downloading(&self) -> Result<String> {
        let body = XmlMap::new();
        post_set(self.conn, "api/online-update/cancel-downloading", &body)
    }

    /// `online-update/upgrade-messagebox`. Upgrade message box.
    pub fn upgrade_messagebox(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/upgrade-messagebox")
    }

    /// Set the upgrade message box.
    pub fn set_upgrade_messagebox(&self, messagebox: &str) -> Result<String> {
        let body: XmlMap = map_of([("messagebox", messagebox.to_string())]);
        post_set(self.conn, "api/online-update/upgrade-messagebox", &body)
    }

    /// `online-update/configuration`. Online update configuration.
    pub fn configuration(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/configuration")
    }

    /// `online-update/autoupdate-config`. Auto-update configuration.
    pub fn autoupdate_config(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/autoupdate-config")
    }

    /// Set the auto-update configuration.
    ///
    /// * `autoupdate` — enable/disable automatic updates.
    pub fn set_autoupdate_config(&self, autoupdate: bool) -> Result<String> {
        let body: XmlMap = map_of([
            (
                "auto_update",
                if autoupdate {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            ("ui_download", "0".to_string()),
        ]);
        post_set(self.conn, "api/online-update/autoupdate-config", &body)
    }

    /// `online-update/redirect_cancel`. Cancel the redirection.
    pub fn redirect_cancel(&self) -> Result<Value> {
        get_value(self.conn, "api/online-update/redirect_cancel")
    }
}
